server/src/main/java/org/elasticsearch/common/util/BytesRefHash.java

```diff
@@ -51,6 +55,41 @@ public BytesRefHash(long capacity, float maxLoadFactor, BigArrays bigArrays) {
         spare = new BytesRef();
     }
 
+    public BytesRefHash(BytesRefArray bytesRefArray, BigArrays bigArrays) {
+        this(bytesRefArray, DEFAULT_MAX_LOAD_FACTOR, bigArrays);
+    }
+
+    public BytesRefHash(BytesRefArray byteRefs, float maxLoadFactor, BigArrays bigArrays) {
+        super(byteRefs.size() + 1, maxLoadFactor, bigArrays);
+        this.bytesRefs = byteRefs;
+
+        boolean success = false;
+        try {
+            // `super` allocates a big array so we have to `close` if we fail here or we'll leak it.
+            hashes = bigArrays.newIntArray(capacity(), false);
+            success = true;
+        } finally {
+            if (false == success) {
+                close();
+            }
+        }
+        spare = new BytesRef();
+
+        // recreate hashes
+        for (int i = 0; i < byteRefs.size(); ++i) {
+            byteRefs.get(i, spare);
+            int code = rehash(spare.hashCode());
+            final long slot = slot(code, mask);
```

Comment:
There's a reset method on this class that we can use to eliminate this code duplication.